js_modules/dagster-ui/packages/ui-core/src/runs/useCursorAccumulatedQuery.tsx

```diff
@@ -1,45 +1,125 @@
+  private onData: (data: DataType[]) => void;
+  private onError?: (error: ErrorType) => void;
+
+  private hasMoreData = true;
+  private dataSoFar: DataType[] = [];
+  private currentCursor: CursorType | undefined = undefined;
+  private fetchPromise?: Promise<void>;
+  private stopped: boolean = false;
+
+  constructor({
+    fetchData,
+    onData,
+    onError,
+  }: {
+    fetchData: FetcherFunction<DataType, CursorType, ErrorType>;
+    onData: (data: DataType[]) => void;
+    onError?: (error: ErrorType) => void;
+  }) {
+    this.fetchData = fetchData;
+    this.onData = onData;
+    this.onError = onError;
+  }
+
+  fetch = async () => {
+    if (this.fetchPromise) {
+      return await this.fetchPromise;
+    }
+    this.fetchPromise = new Promise(async (res) => {
+      // make at least one request
+      this.hasMoreData = true;
+
+      // continue requesting with updated cursors + accumulating data until
+      // stop() is called or hasMore=false.
+      while (this.hasMoreData && !this.stopped) {
+        const {cursor, hasMore, data, error} = await this.fetchData(this.currentCursor);
+        if (this.stopped) {
+          break;
+        }
+        if (error) {
+          this.onError?.(error);
+          break;
+        }
+        this.currentCursor = cursor;
+        this.hasMoreData = hasMore;
+        if (data.length > 0) {
+          this.dataSoFar = this.dataSoFar.concat(data);
+          this.onData(this.dataSoFar);
+        }
+      }
+      res();
+    });
+    const result = await this.fetchPromise!;
+    this.fetchPromise = undefined;
+    return result;
+  };
+
+  stop = () => {
+    this.stopped = true;
+  };
+}
+
+export function useCursorAccumulatedQuery<
+  TQuery,
+  TVars extends OperationVariables & {cursor?: CursorType},
+  DataType,
+  ErrorType = unknown,
+  CursorType = TVars['cursor'],
+>({
+  query,
+  variables,
+  getResult,
+}: {
   query: DocumentNode;
-  skip?: boolean;
   variables: Omit<TVars, 'cursor'>;
-  getResultArray: (result: T | undefined) => U[];
-  getNextFetchState: (result: T | undefined) => FetchState | null;
+  getResult: (responseData: TQuery) => FetchResult<DataType, CursorType, ErrorType>;
 }) {
-  const [fetched, setFetched] = useState<U[] | null>(null);
-  const [fetchState, setFetchState] = useState<FetchState>({hasMore: true, cursor: undefined});
-  const {query, skip, variables, getResultArray, getNextFetchState} = options;
-
-  const queryResult = useQuery<T, TVars>(query, {
-    variables: {...variables, cursor: fetchState.cursor} as TVars,
-    notifyOnNetworkStatusChange: true,
-    skip,
-  });
+  const [fetched, setFetched] = useState<DataType[] | null>(null);
+  const [error, setError] = useState<ErrorType | null>(null);
+  const client = useApolloClient();
+
+  const {stop, fetch} = useMemo(() => {
+    return new AccumulatingDataFetcher({
+      fetchData: async (cursor) => {
+        const resp = await client.query<TQuery, TVars>({
```

Comment:
One question I do have about using client.query this way is if it goes through the apollo cache? The old version I believe does.

Reply:
I believe this stores into the apollo cache but never returns from the cache

Reply:
Ah interesting, I wonder if we should read the cache first?